torch/_inductor/fx_passes/quantization.py

```diff
@@ -1344,38 +1360,80 @@ def _generate_qconv_weight_prepack_patterns(dtype=torch.float32):
     )
 
 
-def _is_valid_dequant_linear_pattern(dtype, input_dim_exceeds_two):
-    def _inner(match):
-        # Check dequant pattern has only 1 user.
-        if input_dim_exceeds_two:
+def _get_linear_node(match, input_dim_exceeds_two, input_contiguous):
+    output_reshape_node = None
+    if input_dim_exceeds_two:
+        if input_contiguous:
             output_reshape_node = match.output_node()
             assert output_reshape_node.target is aten.reshape.default
             linear_node = output_reshape_node.args[0]
         else:
-            linear_node = match.output_node()
+            linear_nodes = filter_nodes(match.nodes, aten.bmm.default)
+            assert len(linear_nodes) == 1
+            linear_node = linear_nodes[0]
+    else:
+        linear_node = match.output_node()
 
-        assert linear_node.target in (aten.addmm.default, aten.mm.default)
-        input_index = 0 if linear_node.target is aten.mm.default else 1
-        assert dtype in [torch.float32, torch.bfloat16]
+    assert linear_node.target in (
+        aten.addmm.default,
+        aten.mm.default,
+        aten.bmm.default,
+    )
+    return linear_node, output_reshape_node
 
-        if input_dim_exceeds_two:
+
+def _get_linear_dq_mul_node(
+    linear_node, input_index, dtype, input_dim_exceeds_two, input_contiguous
+):
+    act_reshape_node = None
+    activation_to_bf16_node = None
+    act_expand_node = None
+    if input_dim_exceeds_two:
+        if input_contiguous:
             act_reshape_node = linear_node.args[input_index]
             assert act_reshape_node.target is aten.reshape.default
-            mul_node = (
-                act_reshape_node.args[0]  # pattern: linear -> reshape -> mul
-                if dtype == torch.float32
-                else act_reshape_node.args[0].args[
-                    0
-                ]  # pattern: linear -> reshape -> to_bf16 -> mul
-            )
+            if dtype == torch.float32:
+                # pattern: linear -> reshape -> mul
+                mul_node = act_reshape_node.args[0]
+            else:
+                # pattern: linear -> reshape -> to_bf16 -> mul
+                activation_to_bf16_node = act_reshape_node.args[0]
+                mul_node = activation_to_bf16_node.args[0]
         else:
-            mul_node = (
-                linear_node.args[input_index]  # pattern: linear -> mul
-                if dtype == torch.float32
-                else linear_node.args[input_index].args[
-                    0
-                ]  # pattern: linear -> to_fb16 -> mul
-            )
+            # bmm pattern decomposed from linear when input dim exceeds 2 and not contiguous
```

Comment:
add comment for the pattern here just like other paths?

Reply:
Thanks and add the comments for the linear patterns to do weight prepack.